Command/ReindexCommand.php

```diff
@@ -84,33 +111,124 @@ public function execute(InputInterface $input, OutputInterface $output)
-            foreach ($checkpoints as $name => $value) {
+        if (count($providerNames) > 0) {
+            foreach ($providerNames as $providerName) {
                 $question = new ConfirmationQuestion(sprintf(
-                    '<question>Checkpoint found for "%s", do you wish to resume from %d?</question> ',
-                    $name, $value
+                    '<question>Provider "%s" did not finish. Do you wish to resume?</question> ',
+                    $providerName
                 ));
-                $response = $questionHelper->ask($input, $output, $question, true);
+                $response = $this->questionHelper->ask($input, $output, $question, true);
 
                 if (false === $response) {
-                    $this->resumeManager->removeCheckpoint($name);
+                    $this->resumeManager->removeCheckpoints($providerName);
                 }
             }
         }
 
-        $event = new IndexRebuildEvent($filter, $purge, $output);
-        $this->eventDispatcher->dispatch(SearchEvents::INDEX_REBUILD, $event);
+        $providerNames = $input->getOption('provider');
+
+        if (empty($providerNames)) {
+            $providers = $this->providerRegistry->getProviders();
+        } else {
+            $providers = [];
+            foreach ($providerNames as $providerName) {
+                $providers[$providerName] = $this->providerRegistry->getProvider($providerName);
+            }
+        }
+
+        foreach ($providers as $providerName => $provider) {
+            $output->writeln(sprintf('<info>provider "</>%s<info>"</>', $providerName));
+            $output->write(PHP_EOL);
+
+            foreach ($provider->getClassFqns() as $classFqn) {
+                $return = $this->reindexClass($output, $provider, $providerName, $classFqn, $batchSize);
+
+                if (false === $return) {
+                    break;
+                }
+            }
+
+            $this->resumeManager->removeCheckpoints($providerName);
+        }
 
-        $output->write(PHP_EOL);
         $output->writeln(sprintf(
-            'Index Rebuild Completed (%ss %sb)',
+            '<info>Index rebuild completed (</>%ss %sb</><info>)</>',
             number_format(microtime(true) - $startTime, 2),
             number_format(memory_get_usage())
         ));
     }
+
+    private function reindexClass(
+        OutputInterface $output,
+        ReIndexProviderInterface $provider,
+        $providerName,
+        $classFqn,
+        $batchSize
+    ) {
+        $count = $provider->getCount($classFqn);
+        $checkpoint = $this->resumeManager->getCheckpoint($providerName, $classFqn);
+        $offset = $checkpoint ?: 0;
+
+        // If the offset is the same as the count then the job was already completed.
+        // If more objects are in the database than before, then we will just index a couple
+        // of more documents than normal.
+        if ($offset === $count) {
```

Comment:
If the checkpoint is a couple of weeks old then they probably shouldn't use it :)